source/slang/syntax.cpp

```diff
@@ -1579,7 +1579,11 @@ void Type::accept(IValVisitor* visitor, void* extra)
                 }
             }
         }
-        return this;
+        RefPtr<DeclaredSubtypeWitness> rs = new DeclaredSubtypeWitness();
```

Comment:
This is fine as it is, but I should take the change here to explain what in the world the `ioDiff` thing is all about.

Basically, the 99% case is that when you call `Substitute()`, nothing actually changes. You have `vector<T,3>` and are asked to substitute `U -> float`, and that doesn't apply so you'd like to return the original type without wasting time on memory allocation, etc.

So if you are looking at `vector<..., ...>` and have a substitution, what are you supposed to do, if you want to avoid wasteful allocation? You could:

1. Just don't worry about it. Recursively substitute all your children, then construct a new type object and return it, even if nothing changed. This wouldn't actually be too bad if/when we actually start caching and re-using types instead of creating them all the time, but for now it would be wasteful.

2. Do a pre-emptive pass to check if any of the variables being substituted "occur" in your child values, and don't bother substituting if this check returns `false`. This could work well, but it required maintaining two very similar code paths (one to check, one to actually substitute), so it is error prone (it also has the problem that a leaf node will get checked once for every ancestor, which is a waste).

3. Go ahead and recursively perform substitution on the children, but keep track of a "did anything change?" flag. If after substituting the children, that flag has been set, then set the flag for whoever called you, and construct a new object to return. If the flag does *not* get set, then just return the original object.

The whole `ioDiff` thing is passing along the "did anything change" flag, so ideally in code like this you would substitute all the children and capture the results into local variables, then check the "diff" flag and do a `return this` early exit if nothing actually changed.

One gotcha there is that if you do this optimization, you don't want to just pass down `ioDiff`, but rather your own `int diff = 0;` variable, since there might be the case where `this` is the third argument to a generic, and you've already had a diff on an earlier argument, so `*ioDiff = 1`, and if you pass it down directly, all your children might mistakenly think there has already been a diff, and react accordingly. Instead, you want to only check if there has been a diff local to `this`, and so a fresh `diff` variable is needed. Then if you find that there *has* been a local diff, you add that to the incoming/outgoing `ioDiff`.

Realistically, this is just a symptom of the fact that we *really* need to have a centralized dictionary of types, and thus guarantee that we don't allocate new objects for structurally identical types (actually, it should apply to all `Val`s). Then this logic would be able to follow the naive approach of constructing everything every time, and it would avoid extra allocations at the cost of a few hash-table lookups.